File: shasi/src/types/service.ts
// Re-export product types to avoid duplication
export type { Product, ProductCategory, ProductFormData } from "./product";

export type CommissionType = 'fixed' | 'percentage';

export interface ServiceCategory {
  id: string;
  name: string;
  description: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface Service {
  id: string;
  name: string;
  category_id: string | null;
  description: string | null;
  duration_minutes: number;
  base_price: number;
  // Handling commission: earned when staff is PIC / mengerjakan tindakan (always)
  doctor_commission_type: CommissionType;
  doctor_commission_value: number;
  therapist_commission_type: CommissionType;
  therapist_commission_value: number;
  // Offering commission: earned only when staff offers the service and patient agrees
  doctor_offering_commission_type?: CommissionType | null;
  doctor_offering_commission_value?: number | null;
  therapist_offering_commission_type?: CommissionType | null;
  therapist_offering_commission_value?: number | null;
  is_active: boolean;
  requires_doctor: boolean;
  created_at: string;
  updated_at: string;
  // Joined data
  category?: ServiceCategory;
}

export interface ServiceFormData {
  name: string;
  category_id?: string;
  description?: string;
  duration_minutes?: number;
  base_price: number;
  // Handling commission
  doctor_commission_type?: CommissionType;
  doctor_commission_value?: number;
  therapist_commission_type?: CommissionType;
  therapist_commission_value?: number;
  // Offering commission
  doctor_offering_commission_type?: CommissionType | null;
  doctor_offering_commission_value?: number | null;
  therapist_offering_commission_type?: CommissionType | null;
  therapist_offering_commission_value?: number | null;
  requires_doctor?: boolean;
}

export interface ServiceConsumable {
  id: string;
  service_id: string;
  product_id: string;
  quantity_used: number;
  created_at: string;
}

export interface StockMovement {
  id: string;
  product_id: string;
  movement_type: 'in' | 'out' | 'adjustment';
  quantity: number;
  reason: string | null;
  reference_id: string | null;
  reference_type: string | null;
  notes: string | null;
  created_by: string | null;
  created_at: string;
}

File: sc-pos-be/internal/models/service.go
package models

import "time"

// Service represents clinic service/treatment
type Service struct {
	ID                       string           `json:"id" db:"id"`
	Name                     string           `json:"name" db:"name"`
	CategoryID               *string          `json:"category_id" db:"category_id"`
	Description              *string          `json:"description" db:"description"`
	DurationMinutes          int              `json:"duration_minutes" db:"duration_minutes"`
	BasePrice                float64          `json:"base_price" db:"base_price"`
	DoctorCommissionType     string           `json:"doctor_commission_type" db:"doctor_commission_type"` // percentage, fixed
	DoctorCommissionValue    float64          `json:"doctor_commission_value" db:"doctor_commission_value"`
	TherapistCommissionType  string           `json:"therapist_commission_type" db:"therapist_commission_type"`
	TherapistCommissionValue float64          `json:"therapist_commission_value" db:"therapist_commission_value"`
	RequiresDoctor           bool             `json:"requires_doctor" db:"requires_doctor"`
	IsActive                 bool             `json:"is_active" db:"is_active"`
	CreatedBy                *string          `json:"created_by,omitempty" db:"created_by"`
	UpdatedBy                *string          `json:"updated_by,omitempty" db:"updated_by"`
	DeletedAt                *time.Time       `json:"deleted_at,omitempty" db:"deleted_at"`
	CreatedAt                time.Time        `json:"created_at" db:"created_at"`
	UpdatedAt                time.Time        `json:"updated_at" db:"updated_at"`
	Category                 *ServiceCategory `json:"category,omitempty" db:"-"`
}

// ServiceCategory represents service categories
type ServiceCategory struct {
	ID          string     `json:"id" db:"id"`
	Name        string     `json:"name" db:"name"`
	Description *string    `json:"description" db:"description"`
	IsActive    bool       `json:"is_active" db:"is_active"`
	CreatedBy   *string    `json:"created_by,omitempty" db:"created_by"`
	UpdatedBy   *string    `json:"updated_by,omitempty" db:"updated_by"`
	DeletedAt   *time.Time `json:"deleted_at,omitempty" db:"deleted_at"`
	CreatedAt   time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at" db:"updated_at"`
}
